library/ssl_tls13_generic.c

```diff
@@ -683,27 +684,48 @@ static int ssl_tls13_validate_certificate(mbedtls_ssl_context *ssl)
 #endif /* MBEDTLS_SSL_CLI_C */
     }
 
-#if defined(MBEDTLS_SSL_SERVER_NAME_INDICATION)
-    if (ssl->handshake->sni_ca_chain != NULL) {
-        ca_chain = ssl->handshake->sni_ca_chain;
-        ca_crl = ssl->handshake->sni_ca_crl;
-    } else
-#endif /* MBEDTLS_SSL_SERVER_NAME_INDICATION */
-    {
-        ca_chain = ssl->conf->ca_chain;
-        ca_crl = ssl->conf->ca_crl;
-    }
-
     /*
      * Main check: verify certificate
      */
-    ret = mbedtls_x509_crt_verify_with_profile(
-        ssl->session_negotiate->peer_cert,
-        ca_chain, ca_crl,
-        ssl->conf->cert_profile,
-        ssl->hostname,
-        &verify_result,
-        ssl->conf->f_vrfy, ssl->conf->p_vrfy);
+#if defined(MBEDTLS_X509_TRUSTED_CERTIFICATE_CALLBACK)
+    if (ssl->conf->f_ca_cb != NULL) {
+        have_ca_chain = 1;
```

Comment:
This looks wrong: as soon as a callback exists, we pretend that there's a CA chain, or more accurately, that the certificate is valid? I think we should only set this if `mbedtls_x509_crt_verify_with_ca_cb` returns 0.

Mind you, checking exactly how `have_ca_chain` is used when there's a callback, maybe my comment above about the name of this variable isn't accurate and the variable should actually be `have_callback_or_valid_ca_chain`, since it's only used to change the error code when `authmode == MBEDTLS_SSL_VERIFY_REQUIRED`. I'm not sure what the right thing is here: it seems to me that some error codes from the callback should be reported as such, e.g. an out-of-memory error. But that would apply to the non-callback error path too. So I guess `ret = MBEDTLS_ERR_SSL_CA_CHAIN_REQUIRED` should only be done when `ret` is certain error values? Or just when `ret == 0`?

Note: I realize the logic is the same in the ≤1.2 code path in `ssl_parse_certificate_verify` in `ssl_tls.c`. I'm equally confused there. Which leads me to:

* Why do we have such similar functions? I'd really prefer to have a single function. It would have avoided missing callback support in 1.3.
* Since we've been happy with the logic and naming in the 1.2 code for years (all the way back to https://github.com/Mbed-TLS/mbedtls/pull/2532), I'm willing to accept keeping the logic and naming here as well. I'm not happy with the legibility, but I'm satisfied that the worse that can happen is returning a less convenient error code sometimes.